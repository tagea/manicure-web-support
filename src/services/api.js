import axios from 'axios';

import { store } from '../../index';

const URL = 'https://manicureapp.com';

const endpoints = {
  auth: `${URL}/login/`,
  authOut: `${URL}/logout/`,
  getUser: (page, size) => `${URL}/users/?page=${page}&page_size=${size}`,
  getSalon: (page, size) => `${URL}/salon/?page=${page}&page_size=${size}`
};

// get and set localStorage
export const stateLocalStorage = token => (
  token ? localStorage.setItem('auth', token) : localStorage.getItem('auth')
);

export const removeLocalStorage = () => {
  const token = stateLocalStorage();
  if (token) {
    localStorage.removeItem('auth');
  }
};

export const axiosInstance = axios.create({ headers: { 'Content-Type': 'application/json' } });

axiosInstance.interceptors.response.use(
  response => response,
  (error) => {
    const originalRequest = error.config;
    if (error.response.status === 401 || error.response.status === 400) {
      store.dispatch({ type: 'UNAUTHENTICATED' });
      return axios(originalRequest);
    }
    return Promise.reject(error);
  }
);
axiosInstance.interceptors.request.use(
  (config) => {
    const token = stateLocalStorage();
    // eslint-disable-next-line no-param-reassign
    config.headers.Authorization = token ? `Token ${token}` : '';
    return config;
  },
  (error) => {
    store.dispatch({ type: 'REQUEST_ERROR' });
    return Promise.reject(error);
  }
);
// check username and password => login
export const getAuthToken = async (username, password) => {
  const res = await axiosInstance.post(endpoints.auth, { username, password });
  return res.data.token;
};
// logout
export const sendAuthToken = async () => {
  const res = await axiosInstance.get(endpoints.authOut);
  return res.data;
};
// get user list
export const fetchUser = async (page, size) => {
  const res = await axiosInstance.get(endpoints.getUser(page, size));
  return res;
};
// get salon list
export const fetchSalon = async (page, size) => {
  const res = await axiosInstance.get(endpoints.getSalon(page, size));
  return res;
};
