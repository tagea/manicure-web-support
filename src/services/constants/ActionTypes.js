export const AUTHENTICATED = 'AUTHENTICATED';
export const AUTHENTICATION_ERROR = 'AUTHENTICATION_ERROR';
export const UNAUTHENTICATION = 'UNAUTHENTICATION';
export const LOGOUT = 'LOGOUT';
export const LOGOUT_ERROR = 'LOGOUT_ERROR';
export const FETCH_SALON_SUCCESS = 'FETCH_SALON_SUCCESS';
export const FETCH_SALON_WAITING = 'FETCH_SALON_WAITING';
export const FETCH_SALON_FAILURE = 'FETCH_SALON_FAILURE';
export const FETCH_USER_SUCCESS = 'FETCH_USER_SUCCESS';
export const FETCH_USER_WAITING = 'FETCH_USER_WAITING';
