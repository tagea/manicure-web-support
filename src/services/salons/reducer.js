import {
  FETCH_SALON_SUCCESS,
  FETCH_SALON_WAITING,
  FETCH_SALON_FAILURE,
  LOGOUT
} from '../constants/ActionTypes';

const initState = {
  salons: [],
  waiting: false,
  sizeOfSalons: 0,
  nextPageSalons: null,
  error: null
};

export default function (state = initState, action) {
  switch (action.type) {
    case FETCH_SALON_SUCCESS:
      return Object.assign({}, state, {
        waiting: false,
        salons: action.payload.salons,
        sizeOfSalons: action.payload.count,
        nextPageSalons: action.payload.pagin.next
      });
    case FETCH_SALON_WAITING:
      return Object.assign({}, state, { waiting: true });
    case FETCH_SALON_FAILURE:
      return { ...state, error: action.payload };
    case LOGOUT:
      return { initState };
    default:
      return state;
  }
}
