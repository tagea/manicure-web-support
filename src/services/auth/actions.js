import {
  AUTHENTICATED,
  AUTHENTICATION_ERROR,
  LOGOUT,
  LOGOUT_ERROR
} from '../constants/ActionTypes';
import { getAuthToken, sendAuthToken, stateLocalStorage, removeLocalStorage } from '../api';

export const signInAction = (username, password) => async (dispatch) => {
  try {
    removeLocalStorage();
    const token = await getAuthToken(username, password);
    dispatch({ type: AUTHENTICATED });
    stateLocalStorage(token); // set
  } catch (error) {
    dispatch({
      type: AUTHENTICATION_ERROR,
      payload: error.message
    });
  }
};

export const signOutAction = () => async (dispatch) => {
  try {
    await sendAuthToken();
    removeLocalStorage();
    dispatch({ type: LOGOUT });
  } catch (error) {
    removeLocalStorage();
    dispatch({
      type: LOGOUT_ERROR,
      payload: error.message
    });
  }
};

// // note for future: add token check // //
export const stayInAction = () => (dispatch) => {
  try {
    const token = stateLocalStorage();
    if (token) {
      dispatch({ type: AUTHENTICATED });
    }
  } catch (error) {
    dispatch({
      type: AUTHENTICATION_ERROR,
      payload: error.message
    });
  }
};
