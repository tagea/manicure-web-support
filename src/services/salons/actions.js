import {
  FETCH_SALON_SUCCESS,
  FETCH_SALON_WAITING,
  FETCH_SALON_FAILURE
} from '../constants/ActionTypes';
import { fetchSalon } from '../api';

export const fetchSalonAction = (page, size) => async (dispatch) => {
  dispatch({ type: FETCH_SALON_WAITING });
  try {
    const response = await fetchSalon(page, size);
    dispatch({
      type: FETCH_SALON_SUCCESS,
      payload: response.data
    });
  } catch (error) {
    dispatch({
      type: FETCH_SALON_FAILURE,
      payload: error.message
    });
  }
};
