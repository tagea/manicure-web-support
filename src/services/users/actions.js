import {
  FETCH_USER_SUCCESS,
  FETCH_USER_WAITING,
  UNAUTHENTICATION
} from '../constants/ActionTypes';
import { fetchUser, removeLocalStorage } from '../api';

export const fetchUserAction = (page, size) => async (dispatch) => {
  dispatch({ type: FETCH_USER_WAITING });
  try {
    const response = await fetchUser(page, size);
    dispatch({
      type: FETCH_USER_SUCCESS,
      payload: response.data
    });
  } catch (error) {
    removeLocalStorage();
    dispatch({
      type: UNAUTHENTICATION,
      payload: error.message
    });
  }
};
