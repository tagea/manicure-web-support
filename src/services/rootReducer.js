import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

import auth from './auth/reducer';
import users from './users/reducer';
import salons from './salons/reducer';

export const rootReducer = combineReducers({
  auth,
  users,
  salons,
  routing: routerReducer
});
