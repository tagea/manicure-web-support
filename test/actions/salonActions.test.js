import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import * as types from '../../src/services/constants/ActionTypes';
import { fetchSalonAction } from '../../src/services/salons/actions';
import { getAuthToken, stateLocalStorage } from '../../src/services/api';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore({});
const auth = { username: 'a', password: 'a' };

beforeEach(() => store.clearActions());

it('Dispatches the correct fetchUserAction and payload', () => (
  store.dispatch(fetchSalonAction()).then(() => (
    expect(store.getActions()).toMatchSnapshot()
  ))
));

it('error fetch user list', () => {
  const expectedActions = [
    {
      type: types.FETCH_SALON_WAITING
    },
    {
      type: types.FETCH_SALON_FAILURE,
      payload: 'Request failed with status code 401'
    }
  ];
  return store.dispatch(fetchSalonAction()).then(() => {
    expect(store.getActions()).toEqual(expectedActions);
  });
});

it('correct fetch user list', async () => {
  // create authenticate to get token
  const token = await getAuthToken(auth.username, auth.password);
  stateLocalStorage(token);
  // check user list
  return store.dispatch(fetchSalonAction(1, 1)).then(() => {
    expect(store.getActions()).toMatchObject([
      { type: 'FETCH_SALON_WAITING' },
      { type: 'FETCH_SALON_SUCCESS' }
    ]);
  });
});
