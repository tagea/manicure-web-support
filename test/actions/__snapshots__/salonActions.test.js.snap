// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dispatches the correct fetchUserAction and payload 1`] = `
Array [
  Object {
    "type": "FETCH_SALON_WAITING",
  },
  Object {
    "payload": "Request failed with status code 401",
    "type": "FETCH_SALON_FAILURE",
  },
]
`;
